gno.land/cmd/gnoland/config_set.go

```diff
@@ -0,0 +1,253 @@
+			ShortUsage: "config set <key> <value>",
+			ShortHelp:  "Edits the Gno node configuration",
+			LongHelp: "Edits the Gno node configuration at the given path " +
+				"by setting the option specified at <key> to the given <value>",
+		},
+		cfg,
+		func(_ context.Context, args []string) error {
+			return execConfigEdit(cfg, io, args)
+		},
+	)
+
+	return cmd
+}
+
+func execConfigEdit(cfg *configCfg, io commands.IO, args []string) error {
+	// Load the config
+	loadedCfg, err := config.LoadConfigFile(cfg.configPath)
+	if err != nil {
+		return fmt.Errorf("unable to load config, %w", err)
+	}
+
+	// Make sure the edit arguments are valid
+	if len(args) != 2 {
+		return errInvalidConfigSetArgs
+	}
+
+	var (
+		key   = args[0]
+		value = args[1]
+	)
+
+	// Update the config field
+	if err := updateConfigField(
+		loadedCfg,
+		key,
+		value,
+	); err != nil {
+		return fmt.Errorf("unable to update config field, %w", err)
+	}
+
+	// Make sure the config is now valid
+	if err := loadedCfg.ValidateBasic(); err != nil {
+		return fmt.Errorf("unable to validate config, %w", err)
+	}
+
+	// Save the config
+	if err := config.WriteConfigFile(cfg.configPath, loadedCfg); err != nil {
+		return fmt.Errorf("unable to save updated config, %w", err)
+	}
+
+	io.Printfln("Updated configuration saved at %s", cfg.configPath)
+
+	return nil
+}
+
+// updateFieldAtPath updates the field at the given path, with the given value
+func updateConfigField(config *config.Config, key, value string) error {
+	// Get the config value using reflect
+	configValue := reflect.ValueOf(config).Elem()
+
+	// Get the value path, with sections separated out by a period
+	path := strings.Split(key, ".")
+
+	field, err := getFieldAtPath(configValue, path)
+	if err != nil {
+		return err
+	}
+
+	// We've reached the actual field, check if it can be updated
+	if !field.CanSet() {
+		return fmt.Errorf("unable to set value")
+	}
+
+	// Convert the value to the field's type
+	v, err := convertStringToType(value, field.Interface())
+	if err != nil {
+		return fmt.Errorf("unable to convert value to field type, %w", err)
+	}
+
+	// Update the field value
+	field.Set(reflect.ValueOf(v))
+
+	return nil
+}
+
+// generateInvalidFieldError generates an invalid field error
+func generateInvalidFieldError(field string, value reflect.Value) error {
+	var (
+		valueType = value.Type()
+		numFields = value.NumField()
+	)
+
+	fields := make([]string, numFields)
+
+	for i := 0; i < numFields; i++ {
+		fields[i] = valueType.Field(i).Name
+	}
+
+	return fmt.Errorf(
+		"field \"%s\", is not a valid configuration key, available keys: %s",
+		field,
+		fields,
+	)
+}
+
+// convertStringToType attempts to convert the given
+// string value to an output type.
+// Because we opted to using reflect instead of a flag-based approach,
+// arguments (always strings) need to be converted to the field's
+// respective type, if possible
+func convertStringToType(value string, outputType any) (any, error) {
```

Comment:
Let me show the problems with this with a test :mage_man::

```go
func Test_convertStringToType(t *testing.T) {
	t.Parallel()

	errContains := func(s string) func(error) {
		return func(err error) {
			assert.ErrorContains(t, err, s)
		}
	}
	tt := []struct {
		input      string
		outputType any
		expected   any
	}{
		{"1", uint8(0), uint8(1)},
		{"-1", uint8(0), errContains("unable to convert to uint8")},
		{"12.34", float32(0), float32(12.34)},
	}

	for _, tc := range tt {
		tc := tc
		t.Run(fmt.Sprintf("%s_%T", tc.input, tc.outputType), func(t *testing.T) {
			t.Parallel()

			res, err := convertStringToType(tc.input, tc.outputType)
			if checker, ok := tc.expected.(func(error)); ok {
				checker(err)
			} else {
				assert.Equal(t, tc.expected, res)
			}
		})
	}
}
```

All of these fail:

```
=== NAME  Test_convertStringToType/12.34_float32
    config_set_test.go:929:
                Error Trace:    /home/howl/oc/gno/gno.land/cmd/gnoland/config_set_test.go:929
                Error:          Not equal:
                                expected: float32(12.34)
                                actual  : float64(12.34000015258789)
                Test:           Test_convertStringToType/12.34_float32
=== NAME  Test_convertStringToType/1_uint8
    config_set_test.go:929:
                Error Trace:    /home/howl/oc/gno/gno.land/cmd/gnoland/config_set_test.go:929
                Error:          Not equal:
                                expected: uint8(0x1)
                                actual  : int64(1)
                Test:           Test_convertStringToType/1_uint8
=== NAME  Test_convertStringToType
    config_set_test.go:907:
                Error Trace:    /home/howl/oc/gno/gno.land/cmd/gnoland/config_set_test.go:907
                                                        /home/howl/oc/gno/gno.land/cmd/gnoland/config_set_test.go:927
                Error:          An error is expected but got nil.
                Test:           Test_convertStringToType
```

(The float test would have to be changed anyway due to the usual floating point shenanigans; my point is that the type is different)

There are a number of bugs I spotted when reading this:

- parseUint uses strconv.ParseInt instead of strconv.ParseUint
- The resulting value is not changed to the appropriate type (resulting in the above test cases, where the results are different)
- uint case makes the assumption that the architecture is 64-bit, and in fact you avoided this problem for int by splitting out its case from int64

Though, more generally speaking, I made my original comment instead of these because after having written too many parsers than I'd like to admit, my take is that we should re-use the parsers that exist if they do the same dull, boring thing as we are so we don't have to wrangle with the fine details all over. These are the ones I've noted, there may be more but honestly I'd rather avoid seeing a single bug report because we didn't handle an edge case in this parser.

My proposal for this one is to change the input to be `reflect.Value` and turn this into an actual "unmarshaler" (so, placing the value directly in the struct field, rather than creating one anew and "putting it in" afterwards). For the majority of "dull" values (all integers, floats, bools) I suggest you do this:

```
json.Unmarshal([]byte(value), dstValue.Addr().Interface())
```

(assume `outputType any` -> `dstValue reflect.Value`)

Reply:
Thank you for the detailed explanation 🙏 

I've applied your suggestions, and I actually like the approach of giving a value that will be modified after execution:

[c09b7ec](https://github.com/gnolang/gno/pull/1605/commits/c09b7ec2a9b0e684c5eee7e65151d530930a3b92)